fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,527 @@
+    cached_data: CachedData,
+    /// This is a in memory view of the committed parent finality. We need this as a starting point
+    /// for populating the cache
+    last_committed_finality: TVar<Option<IPCParentFinality>>,
+    /// The ipc agent proxy that works as a back up if cache miss
+    agent: Arc<T>,
+}
+
+/// Tracks the data from the parent
+#[derive(Clone)]
+struct CachedData {
+    height_data: TVar<SequentialKeyCache<BlockHeight, ParentViewPayload>>,
+}
+
+/// Exponential backoff for futures
+macro_rules! retry {
+    ($wait:expr, $retires:expr, $f:expr) => {{
+        let mut retries = $retires;
+        let mut wait = $wait;
+
+        loop {
+            let res = $f;
+            if let Err(e) = &res {
+                tracing::warn!(
+                    "cannot query ipc agent due to: {e}, retires: {retries}, wait: {wait}"
+                );
+                if retries > 0 {
+                    retries -= 1;
+
+                    let to_sleep = Duration::from_secs(wait);
+                    tokio::time::sleep(to_sleep).await;
+
+                    wait *= 2;
+                    continue;
+                }
+            }
+
+            break res;
+        }
+    }};
+}
+
+#[async_trait::async_trait]
+impl<T: ParentQueryProxy + Send + Sync + 'static> ParentViewProvider for CachedFinalityProvider<T> {
+    /// Should always return the validator set, only when ipc agent is down after exponeitial
+    /// retries
+    async fn validator_changes(
+        &self,
+        height: BlockHeight,
+    ) -> anyhow::Result<Vec<StakingChangeRequest>> {
+        let r = atomically(|| self.cached_data.validator_changes(height)).await;
+        if let Some(v) = r {
+            return Ok(v);
+        }
+
+        retry!(
+            self.config.exponential_back_off_secs,
+            self.config.exponential_retry_limit,
+            self.agent.get_validator_changes(height).await
+        )
+    }
+
+    /// Should always return the top down messages, only when ipc agent is down after exponential
+    /// retries
+    async fn top_down_msgs(&self, height: BlockHeight) -> anyhow::Result<Vec<CrossMsg>> {
+        let r = atomically(|| self.cached_data.top_down_msgs_at_height(height)).await;
+        if let Some(v) = r {
+            return Ok(v);
+        }
+
+        retry!(
+            self.config.exponential_back_off_secs,
+            self.config.exponential_retry_limit,
+            self.agent.get_top_down_msgs(height, height).await
+        )
+    }
+}
+
+impl<T: ParentQueryProxy + Send + Sync + 'static> ParentFinalityProvider
+    for CachedFinalityProvider<T>
+{
+    fn next_proposal(&self) -> Stm<Option<IPCParentFinality>> {
+        let height = if let Some(h) = self.cached_data.latest_height()? {
+            h
+        } else {
+            return Ok(None);
+        };
+
+        // safe to unwrap as latest height exists
+        let block_hash = self.cached_data.block_hash(height)?.unwrap();
```

Comment:
It would be safer to bundle the height and the block into a single `atomically`, to avoid all possibility that between the two calls something else removes that block. It is what it's about, after all.